apps/prairielearn/elements/pl-external-grader-results/pl-external-grader-results.py

```diff
@@ -39,129 +41,140 @@ def ansi_to_html(output):
-                        test["has_images"] = len(test["images"]) > 0
-                        tests.append(test)
-
-                    html_params["tests"] = tests
-
-        with open("pl-external-grader-results.mustache", "r", encoding="utf-8") as f:
-            html = chevron.render(f, html_params).strip()
-    else:
-        html = ""
-
-    return html
+def round_value(value: float, digits: int = 2) -> str:
+    """
+    Round the given value to the specified precision; default is 2 digits.
+
+    Remove trailing 0s and '.'s, e.g., convert "1.00" to "1".
+    """
+    return f"{value:.{digits}f}".rstrip("0").rstrip(".")
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
+    # Early-exit if not the submission panel
+    if data["panel"] != "submission":
+        return ""
+
+    feedback = data["feedback"]
+    results = feedback.get("results", None)
+
+    format_errors = data.get("format_errors", {})
+    grader_format_errors = format_errors.get("_external_grader", [])
+
+    grading_succeeded = bool(feedback.get("succeeded", None))
+
+    html_params = {
+        "submission": True,
+        "graded": bool(feedback),
+        "gradable": results["gradable"]
+        if (results is not None and "gradable" in results)
+        else True,
+        "grading_succeeded": grading_succeeded,
+        "format_errors": grader_format_errors,
+        "has_format_errors": len(grader_format_errors) > 0,
+        "uuid": pl.get_uuid(),
+    }
+
+    if grading_succeeded and results:
+        html_params["succeeded"] = bool(results.get("succeeded", None))
+        html_params["score"] = round_value(results.get("score", 0) * 100)
+        html_params["achieved_max_points"] = results.get("score", 0) >= 1.0
+        html_params["results_color"] = (
+            PLColor("correct_green")
+            if (results.get("score", 0) >= 1.0)
+            else PLColor("incorrect_red")
+        )
+
+        html_params["message"] = ansi_to_html(results.get("message", None))
+        html_params["has_message"] = bool(html_params["message"])
+
+        html_params["output"] = ansi_to_html(results.get("output", None))
+        html_params["has_output"] = bool(html_params["output"])
+
+        html_params["images"] = results.get("images", [])
+        html_params["has_images"] = bool(html_params["images"])
+
+        html_params["has_message_or_output_or_image"] = (
+            html_params["has_message"]
+            or html_params["has_output"]
+            or html_params["has_images"]
+        )
+
+        results_tests = results.get("tests", None)
+        html_params["has_tests"] = bool(results_tests)
+
+        if results_tests:
+            # Let's not assume that people give us a valid array of tests
+            # If any test is missing either points or max_points, we'll
+            # disable detailed scores for all questions
+            tests_missing_points = any(
+                [
+                    test.get("points", None) is None
```

Comment:
```suggestion
                    "points" in test
```
If `"points"` is guaranteed to be an integer (i.e. not a key for `None`), we can just check if the key is present.

Reply:
Shouldn't it be `"points" not in test or test["points"] is None`? I think in the suggestion the logic is inverted.

Reply:
Ah yes sorry, you're right, should be inverted